services/audit-service/src/main/java/datawave/microservice/audit/controller/AuditController.java

```diff
@@ -17,34 +27,131 @@
+import static datawave.webservice.common.audit.AuditParameters.QUERY_AUDIT_TYPE;
+import static datawave.webservice.common.audit.AuditParameters.QUERY_AUTHORIZATIONS;
+import static datawave.webservice.common.audit.AuditParameters.QUERY_DATE;
+import static datawave.webservice.common.audit.AuditParameters.QUERY_LOGIC_CLASS;
+import static datawave.webservice.common.audit.AuditParameters.QUERY_SECURITY_MARKING_COLVIZ;
+import static datawave.webservice.common.audit.AuditParameters.QUERY_SELECTORS;
+import static datawave.webservice.common.audit.AuditParameters.QUERY_STRING;
+import static datawave.webservice.common.audit.AuditParameters.USER_DN;
+
+/**
+ * The AuditController presents the REST endpoints for the audit service.
+ * <p>
+ * Before returning success to the caller, the audit controller will verify that the audit message was successfully passed to our messaging infrastructure.
+ * Also, if configured, the audit controller will verify that the message passing infrastructure is healthy before returning successfully to the user. If the
+ * message passing infrastructure is unhealthy, or if we can't verify that the message was successfully passed to our messaging infrastructure, a 500 Internal
+ * Server Error will be returned to the caller.
+ */
 @RestController
 @RequestMapping(path = "/v1", produces = MediaType.APPLICATION_JSON_VALUE)
 public class AuditController {
     private final Logger log = LoggerFactory.getLogger(this.getClass());
     
+    private final AuditProperties auditProperties;
+    
     private AuditParameters restAuditParams;
     
     private MessageChannel messageChannel;
     
-    public AuditController(@Qualifier("restAuditParams") AuditParameters restAuditParams,
+    @Autowired(required = false)
+    private HealthChecker healthChecker;
+    
+    private static final Map<String,CountDownLatch> correlationLatchMap = new ConcurrentHashMap<>();
+    
+    public AuditController(AuditProperties auditProperties, @Qualifier("restAuditParams") AuditParameters restAuditParams,
                     @Qualifier(AuditServiceConfig.AuditSourceBinding.NAME) MessageChannel messageChannel) {
+        this.auditProperties = auditProperties;
         this.restAuditParams = restAuditParams;
         this.messageChannel = messageChannel;
     }
     
+    /**
+     * Receives producer confirm acks, and disengages the latch associated with the given correlation ID.
+     * 
+     * @param message
+     *            the confirmation ack message
+     */
+    @ConditionalOnProperty(value = "audit.confirmAckEnabled", havingValue = "true", matchIfMissing = true)
+    @ServiceActivator(inputChannel = CONFIRM_ACK_CHANNEL)
+    public void processConfirmAck(Message<?> message) {
+        String correlationId = null;
+        if (message.getHeaders().containsKey(IntegrationMessageHeaderAccessor.CORRELATION_ID)) {
+            Object headerObj = message.getHeaders().get(IntegrationMessageHeaderAccessor.CORRELATION_ID);
```

Comment:
Is getHeaders() not a normal map? Since you're null-checking the value, you don't need to test if it's in the map. Just get it and you'll either get the value or null.